Bluetooth/src/main/java/com/trifork/bluetooth/le/BluetoothManager.kt

```diff
@@ -287,16 +288,14 @@ class BluetoothManager(private val context: Context, private val configuration:
                         }
                     }
                     status == GATT_ERROR || newState == GATT_ERROR -> {
-                        connectedGatt.remove(connection)?.close()
-                        Thread.sleep(200L)
-                        retries += 1
-                        w("Changed connection state from: $status to $newState for device: ${connection.device.address}. Retry attempt #$retries...")
-                        val retryGatt = device.connectGatt(context, false, this)
-                        if (retryGatt != null) {
-                            communicationCallbacks[connection] = this
-                            connectedGatt[connection] = retryGatt
-                        } else {
-                            connectCallback.on(Failure.Bluetooth(Reason.OperationFailed))
+                        val that = this
+                        mainScope.launch {
+                            retries += 1
+                            w("Changed connection state from: $status to $newState for device: ${connection.device.address}. Retry attempt #$retries...")
+
+                            connectedGatt.remove(connection)?.disconnect()
```

Comment:
The connectedGatt is a resource that is accessed  by asynchronous tasks, I think the variable should be synchronised to avoid race conditions(This actually could lead to non normal  behaviour depending on the phone speed).

Reply:
Made sure connectedGatt is only manipulated on the main thread.